sql: refactor pg_builtin to use actual grant options

ecwall opened the pull request:
refs https://github.com/cockroachdb/cockroach/issues/73129

The builtins has_table_privilege, has_column_privilege,
has_any_column_privilege now use privileges.Priv.GrantOption instead 
of privileges.Kind.GRANT.

Refactor builtins.priv -> privilege.Privilege.
    
Replace privilege.Kind with privilege.Privilege in functions that need
access to privilege.Privilege.GrantOption.

ecwall committed:
sql: refactor pg_builtin to use actual grant options

Refactor builtins.priv -> privilege.Privilege.

Replace privilege.Kind with privilege.Privilege in functions that need
access to privilege.Privilege.GrantOption.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/74172)
<!-- Reviewable:end -->

ecwall commented on pkg/sql/privilege/privilege.go:
```diff
@@ -45,6 +45,48 @@ const (
+//
+//  res = check(SELECT) || (check(UPDATE) && check(GRANT)) || check(DELETE)
+//
+type Priv struct {
```
Made this public and moved from `builtins` to `privilege` because I thought it made more sense for function signatures that were using `privilege.Kind` to use `privilege.Priv` instead of `builtins.Priv`.

ecwall commented on pkg/sql/resolver.go:
```diff
@@ -314,6 +314,14 @@ func (p *planner) HasPrivilege(
 		return false, err
 	}
 
+	err = p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, priv.GrantOption)
```
Is this correct?

rafiss replied:
i think instead it should look like:
```
if priv.GrantOption {
  err = p.CheckGrantOptionsForUser(..., true /* isGrant */)
  // rest of the error checking logic
}
```

also, it might need to be done as part of (or more similarly to) hasPrivilegeFunc below.

best way to make sure it's working right is to run the logic tests:

```
make testbaselogic FILES='privilege_builtins'
```
(could also be with bazel)

ecwall replied:
Also, does this need to be wrapped in a version gate to only work in the new version (reverse of this https://github.com/cockroachdb/cockroach/pull/74172#discussion_r773466386)?

ecwall commented on pkg/sql/sem/builtins/pg_builtins.go:
```diff
@@ -583,22 +541,22 @@ func runPrivilegeChecks(
-		if priv.grantOption {
-			// grantOption is set, so AND the result with check(GRANT).
-			d, err = check(privilege.GRANT)
+		if oldVersion { //todo
```
What is the best way to get this context from here?
`p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.DefaultPrivileges)`

rafiss replied:
hm yeah, we don't have a `planner` here. let me think about it. perhaps the version gate needs to be done deeper in the stack?

rafiss reviewed (changes requested):
nit on commit message title -- it shouldn't include the issue number

ecwall commented on pkg/sql/sem/builtins/pg_builtins.go:
```diff
@@ -1488,19 +1446,19 @@ SELECT description
-				"UPDATE WITH GRANT OPTION":     {privilege.UPDATE, true},
-				"REFERENCES":                   {privilege.SELECT, false},
-				"REFERENCES WITH GRANT OPTION": {privilege.SELECT, true},
+				"SELECT":                       {Kind: privilege.SELECT},
```
These lines were generating errors like this during `./dev build`
```
compilepkg: nogo: errors found by nogo during build-time code analysis:
/var/folders/hc/gng26svn1yq13l017705xqjr0000gq/T/rules_go_work-3414600577/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/pg_builtins.go:559:19: github.com/cockroachdb/cockroach/pkg/sql/privilege.Priv composite literal uses unkeyed fields (composites)
```

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[-- commits, line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaLAdCAvJ8TXd5N3wJ):*
The sentence here would be a good fit for the main commit message body. The release notes are written for users -- so the release note should not include internal implementation details like the names of structs or variables.
___
*[pkg/sql/resolver.go, line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaVj547ZklGJzlPNtr:-MsaVj55EKSQZaOwDkZ4:bykjyqs) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L306)):*
> ```Go
> 	specifier tree.HasPrivilegeSpecifier,
> 	user security.SQLUsername,
> 	priv privilege.Priv,
> ```

i think this refactor will end up being pretty invasive. it will need a lot of the usages of this function to be updated. that might be OK -- but i think that refactor should be done separately from this PR. so i prefer to keep this as `privilege.Kind` for now.
___
*[pkg/sql/privilege/privilege.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MrTrB8ll3HL7zQ0tR2t-r1-85:-MsaUhWI61Dc3uGMM-MW:b30lhst) ([raw file](https://github.com/cockroachdb/cockroach/blob/e071ab64fcf0b56536b0c24d7f3ef71fc6eceb64/pkg/sql/privilege/privilege.go#L85)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Made this public and moved from `builtins` to `privilege` because I thought it made more sense for function signatures that were using `privilege.Kind` to use `privilege.Priv` instead of `builtins.Priv`.
</blockquote></details>

this is a good point to think about. i feel like as long as this struct is only used within the builtins package, it would be better to keep it there and keep it non-visible

`builtins.priv` makes sense to me -- it's the representation of a privilege that is relevant for builtin functions. (also the comment on it is very specific for builtin functions too)
___
*[pkg/sql/privilege/privilege.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaV5zD3N6GujBkQcLv:-MsaV5zD3N6GujBkQcLw:b-6xcpta) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/privilege/privilege.go#L59)):*
> ```Go
> // construction, there is no need for a separate GRANT privilege.
> //
> // In CockroachDB, there exists a distinct GRANT privilege and no concept of
> ```

this comment paragraph and the next one need to be updated. it should mention that we do match the postgres "grant option" behavior
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MrTrB8ll3HL7zQ0tR2v-r1-557:-MsaLwnC4cVePwQ28hGw:bym2djv) ([raw file](https://github.com/cockroachdb/cockroach/blob/e071ab64fcf0b56536b0c24d7f3ef71fc6eceb64/pkg/sql/sem/builtins/pg_builtins.go#L557)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

hm yeah, we don't have a `planner` here. let me think about it. perhaps the version gate needs to be done deeper in the stack?
</blockquote></details>

i think the best way is to not do the version gate here at all.

instead, we can remove the check for GRANT in this part of the code. then we add the version gate to the `hasPrivilege` function in pg_builtins.go. on the old version, if `priv.GrantOption` is true, that function should check for the GRANT privilege, but on the new version it should check for the grant option.


<!-- Sent from Reviewable.io -->

rafiss reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1449 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MrTrB8ll3HL7zQ0tR2v-r2-1449:-MsaWLc_EeUrSABwAVMX:bpl591g) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/sem/builtins/pg_builtins.go#L1449)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

These lines were generating errors like this during `./dev build`
```
compilepkg: nogo: errors found by nogo during build-time code analysis:
/var/folders/hc/gng26svn1yq13l017705xqjr0000gq/T/rules_go_work-3414600577/github.com/cockroachdb/cockroach/pkg/sql/sem/builtins/pg_builtins.go:559:19: github.com/cockroachdb/cockroach/pkg/sql/privilege.Priv composite literal uses unkeyed fields (composites)
```
</blockquote></details>

yeah, i think since the `Priv` struct was made public, the linter now requires the field names to be spelled out fully


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[-- commits, line 4 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaLAdCAvJ8TXd5N3wJ):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

The sentence here would be a good fit for the main commit message body. The release notes are written for users -- so the release note should not include internal implementation details like the names of structs or variables.
</blockquote></details>

Ok, good to know.
___
*[pkg/sql/resolver.go, line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaVj547ZklGJzlPNtr:-Msavx59EOt1OmGGUWug:bk4mlfc) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L306)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this refactor will end up being pretty invasive. it will need a lot of the usages of this function to be updated. that might be OK -- but i think that refactor should be done separately from this PR. so i prefer to keep this as `privilege.Kind` for now.
</blockquote></details>

The privilege.Kind -> privilege.Priv refactor is required for this check https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L325 (not sure how to link in reviewable).
___
*[pkg/sql/privilege/privilege.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MrTrB8ll3HL7zQ0tR2t-r1-85:-MsawxAmADReWzFT_mYg:b-n4dczh) ([raw file](https://github.com/cockroachdb/cockroach/blob/e071ab64fcf0b56536b0c24d7f3ef71fc6eceb64/pkg/sql/privilege/privilege.go#L85)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this is a good point to think about. i feel like as long as this struct is only used within the builtins package, it would be better to keep it there and keep it non-visible

`builtins.priv` makes sense to me -- it's the representation of a privilege that is relevant for builtin functions. (also the comment on it is very specific for builtin functions too)
</blockquote></details>

It is passed down and eventually used in resolver.go (https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L325).  Is there a different place I should be checking priv.GrantOption?


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/resolver.go, line 306 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaVj547ZklGJzlPNtr:-MscsBKh1QOWSdyNoIip:b-ig8c7g) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L306)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

The privilege.Kind -> privilege.Priv refactor is required for this check https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L325 (not sure how to link in reviewable).
</blockquote></details>


i was suggesting to do the check in hasPrivilege in pg_builtins.go , and undo all the changes in resolver.go. this would also allow the `priv` struct to remain private in the `builtins` package. but i think what you're doing here is better. i was also wrong about this refactor being invasive; the function is only used in one place. (i remember now that i checked this when writing up the github issue)
___
*[pkg/sql/resolver.go, line 326 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MscwRbHCXJm2HTDyXTz:-MscwRbHCXJm2HTDyXU-:b-xl2l7d) ([raw file](https://github.com/cockroachdb/cockroach/blob/c51f456f571b6a08a0586bc666f797b7dc61e59c/pkg/sql/resolver.go#L326)):*
> ```Go
> 		}
> 		/*if priv.GrantOption { todo
> 			err = p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, true)
> ```

so here we could have:

```
if priv.GrantOption {
  if !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.ValidateGrantOption) {
    ctx.Planner.CheckPrivilegeForUser( ... GRANT)
  } else {
    ctx.Planner.CheckGrantOptionsForUser( ... )
 }
```

(of course, with the correct error handling)
___
*[pkg/sql/privilege/privilege.go, line 85 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MrTrB8ll3HL7zQ0tR2t-r1-85:-MscsUi8CwQGgqQ3rn3n:bqj9trm) ([raw file](https://github.com/cockroachdb/cockroach/blob/e071ab64fcf0b56536b0c24d7f3ef71fc6eceb64/pkg/sql/privilege/privilege.go#L85)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It is passed down and eventually used in resolver.go (https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/resolver.go#L325).  Is there a different place I should be checking priv.GrantOption?
</blockquote></details>

my suggestion was to call `CheckGrantOptions` from the `HasPrivilege` function of builtins.go -- but after thinking a bit more i think that maybe what you're doing now is better
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 557 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MrTrB8ll3HL7zQ0tR2v-r1-557:-MscwU8D9MuLAnqwigc2:bvcp44) ([raw file](https://github.com/cockroachdb/cockroach/blob/e071ab64fcf0b56536b0c24d7f3ef71fc6eceb64/pkg/sql/sem/builtins/pg_builtins.go#L557)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think the best way is to not do the version gate here at all.

instead, we can remove the check for GRANT in this part of the code. then we add the version gate to the `hasPrivilege` function in pg_builtins.go. on the old version, if `priv.GrantOption` is true, that function should check for the GRANT privilege, but on the new version it should check for the grant option.
</blockquote></details>

my bad; actually let's do the version gate in `planner.HasPrivilege`. see my comment there.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall, @rafiss, and @stevendanna)

___
*[pkg/sql/privilege/privilege.go, line 50 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-Msk9-HM7gwhW9OuYUEM:-Msk9-HN89lYIPKpB2M4:b-ddubvt) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad23dac5d29d9ca04416946d7d6e93dc8b1cb27/pkg/sql/privilege/privilege.go#L50)):*
> ```Go
> // Privilege represents a privilege parsed from an Access Privilege Inquiry
> // Function's privilege string argument.
> type Privilege struct {
> ```

It seems like postgres doesn't have a term for a container concept like this around the privilege (what we call Kind) and its grant option so I think calling it Privilege is fine unless you have a better name.

https://www.postgresql.org/docs/13/ddl-priv.html


<!-- Sent from Reviewable.io -->

ecwall committed:
sql: refactor pg_builtin to use actual grant options

The builtins has_table_privilege, has_column_privilege,
has_any_column_privilege now use privileges.Priv.GrantOption instead
of privileges.Kind.GRANT.

Release note: None

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

*[a discussion](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslLJVh3SMaSmeVRi-F:-MslLJVh3SMaSmeVRi-G:bbrphq7) (no related file):*
Is this function still needed? Or can it be replaced by calling `check` inside `runPrivilegeChecks`?
```
func runSinglePrivilegeCheck(
	priv privilege.Privilege, check func(privilege.Privilege) (tree.Datum, error),
) (tree.Datum, error) {
	d, err := check(priv)
	if err != nil {
		return nil, err
	}
	switch d {
	case tree.DBoolFalse, tree.DNull:
	case tree.DBoolTrue:
		if priv.GrantOption {
			// GrantOption is set, so AND the result with check(GRANT).
			d, err = check(privilege.Privilege{Kind: privilege.GRANT})
			if err != nil {
				return nil, err
			}
			switch d {
			case tree.DBoolFalse, tree.DBoolTrue, tree.DNull:
			default:
				return nil, errors.AssertionFailedf(
					"unexpected privilege check result %v", d)
			}
		}
	default:
		return nil, errors.AssertionFailedf(
			"unexpected privilege check result %v", d)
	}
	return d, nil
}
```

---


<!-- Sent from Reviewable.io -->

rafiss reviewed (changes requested):
could you update the PR description so it summarizes the commits? or you could just copy the commit messages into the PR description

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-:-MslNaTW1xJ8-pWIkOtx:bw7d6gh)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[-- commits, line 2 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslM8dZAAtS_Y6RUpK8):*
the first commit message title should be updated -- this commit is just a refactor
___
*[pkg/sql/resolver.go, line 324 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMc_K79QKOQ-kz1Gd:-MslMc_L38Knf3_Wd6NC:b2e1mwl) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/resolver.go#L324)):*
> ```Go
> 					err = p.CheckPrivilegeForUser(ctx, desc, privilege.GRANT, user)
> 				} else {
> 					err = p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, true)
> ```

nit: when passing in constant values to a function, the parameter name should be documented, like this:
```
p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, true /* isGrant */)
```
___
*[pkg/sql/resolver.go, line 329 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMWCS4_w4pHWdYS68:-MslMWCS4_w4pHWdYS69:bhfwc92) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/resolver.go#L329)):*
> ```Go
> 		}
> 		if err != nil {
> 			if pgerror.GetPGCode(err) == pgcode.InsufficientPrivilege {
> ```

we'll need to be a bit careful with the error handling here. i believe a different code is returned by CheckGrantOptionsForUser
___
*[pkg/sql/privilege/privilege.go, line 50 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-Msk9-HM7gwhW9OuYUEM:-MslNUnH3MLkvLLiZoNQ:bq0ncok) ([raw file](https://github.com/cockroachdb/cockroach/blob/9ad23dac5d29d9ca04416946d7d6e93dc8b1cb27/pkg/sql/privilege/privilege.go#L50)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

It seems like postgres doesn't have a term for a container concept like this around the privilege (what we call Kind) and its grant option so I think calling it Privilege is fine unless you have a better name.

https://www.postgresql.org/docs/13/ddl-priv.html
</blockquote></details>

sounds good
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 559 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMNQn1mTs4JYsksjv:-MslMNQn1mTs4JYsksjw:b-bed5vd) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/sem/builtins/pg_builtins.go#L559)):*
> ```Go
> 		if priv.GrantOption {
> 			// GrantOption is set, so AND the result with check(GRANT).
> 			d, err = check(privilege.Privilege{Kind: privilege.GRANT})
> ```

this should be removed in the 2nd commit, since this is now checked in `HasPrivilege`


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/resolver.go, line 329 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMWCS4_w4pHWdYS68:-Mt4iE0o1dLeyZTHFUvo:b-3tn3fm) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/resolver.go#L329)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

we'll need to be a bit careful with the error handling here. i believe a different code is returned by CheckGrantOptionsForUser
</blockquote></details>

Changed it to WarningPrivilegeNotGranted for that case.
___
*[pkg/sql/privilege/privilege.go, line 59 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MsaV5zD3N6GujBkQcLv:-Mt4iB1n6VLkeE_J3W2m:b-bu91xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/dd8a8d8a43b1e7fc4ec74f323afc75259f32204d/pkg/sql/privilege/privilege.go#L59)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this comment paragraph and the next one need to be updated. it should mention that we do match the postgres "grant option" behavior
</blockquote></details>

Updated
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 559 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMNQn1mTs4JYsksjv:-Mt4i0c_-ys1tTL6_WfT:brgp810) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/sem/builtins/pg_builtins.go#L559)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

this should be removed in the 2nd commit, since this is now checked in `HasPrivilege`
</blockquote></details>

I added a todo above with details.


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/resolver.go, line 324 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMc_K79QKOQ-kz1Gd:-Mt4iIz1BluMyptOzYT8:b-bu91xm) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/resolver.go#L324)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: when passing in constant values to a function, the parameter name should be documented, like this:
```
p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, true /* isGrant */)
```
</blockquote></details>

Updated


<!-- Sent from Reviewable.io -->

ecwall reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)

___
*[pkg/sql/resolver.go, line 329 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-MslMWCS4_w4pHWdYS68:-Mt8Sjz5EbXYkc0Pi_k7:b-uudacx) ([raw file](https://github.com/cockroachdb/cockroach/blob/acc4b1d140a1c9c84caab9e5335203e4d16beceb/pkg/sql/resolver.go#L329)):*
<details><summary><i>Previously, ecwall (Evan Wall) wrote…</i></summary><blockquote>

Changed it to WarningPrivilegeNotGranted for that case.
</blockquote></details>

Actually it looks like that was causing tests to fail so I changed it back.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
lgtm!

> Reviewed 1 of 6 files at r4, 1 of 1 files at r5, 1 of 4 files at r7, 1 of 3 files at r10, 2 of 2 files at r11, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74172#-:-MtDqt738OoW_ALC8G2S:b-7nbn20)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ecwall and @rafiss)



<!-- Sent from Reviewable.io -->
